### DISCUSSION TITLE
Sending and Receiving Cloud Events with Kafka

### DISCUSSION THREAD

**Discussion by giscus** [quarkusio/quarkus#33017] [Quarkus Blog/Website]
# Sending and Receiving Cloud Events with Kafka

Quarkus: Supersonic Subatomic Java

https://quarkus.io/blog/kafka-cloud-events/

**Comment by FabioBentoLuiz** [comment]
Hi, I am trying to customize the Cloud Events outgoing messages as described on the section [Sending Cloud Events on Kafka](https://quarkus.io/blog/kafka-cloud-events/#sending-cloud-events-on-kafka) but it fails with the folowing error:

SRMSG00053: Invalid method annotated with @Outgoing: com.example.ExampleProducer#toCloudEvents - no parameters expected

The error message is clear and after digging deeper what worked for me was de following:

```
    @Inject
    @Channel("my-channel")
    Emitter<MyCustomEvent> emitter;

    public void produce(MyCustomEvent event) {
        var metadata = OutgoingCloudEventMetadata.builder()
                .withId(event.getMessageId())
                .withType(event.getClass().getTypeName())
                .build();
        emitter.send(Message.of(event).addMetadata(metadata));
    }
```

As the same example is on the [Smallrye](https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/3.8/kafka/kafka.html#_sending_cloud_events) docs I would like to check if I am missing something or is it something wrong with the docs.

**Comment by jayachristina** [comment]
did you manage to find a solution for this please?

**Reply by FabioBentoLuiz** [reply]
The code snippet in my post is working for me.

**Comment by jayachristina** [comment]
oh yes.. sorry  dont know what I was thinking
but this doesnt work either for me.. only my original payload is getting posted  to kafka.. there is no cloud event
i think I am missing something very basic..

what exactly are you @override please? 

Do you have this code on git somewhere please?

**Reply by FabioBentoLuiz** [reply]
> what exactly are you @OverRide please?

This class is implementing an interface but it is totally irrelevant here.

> Do you have this code on git somewhere please?

The project is not open source, but if you are able to assemble a minimum project reproducing your problem I could have a look.

**Comment by jayachristina** [comment]
https://github.com/rh-soln-pattern-sentiment-analysis/product-reviews/tree/cloudevents

here is it
Thanks @FabioBentoLuiz 

Post any JSON to  ProductReviewResource.java 
http://localhost:8080/review/submit 

and KafkaService writes to a kafka topic called globex.reviews

**Reply by FabioBentoLuiz** [reply]
It was missing some configuration on your application.properties. Check out the PR.

**Reply by jayachristina** [reply]
phew!! Case of the missing config!!
Thanks soo much Fabio!
I was missing `mp.messaging.outgoing.product-reviews.cloud-events-mode=structured`

Appreciate your help :)))))))
